lib/bolt/target.rb

```diff
@@ -1,8 +1,143 @@
+                                target_alias = nil,
+                                config = nil,
+                                facts = nil,
+                                vars = nil,
+                                features = nil)
+      inventory = Puppet.lookup(:bolt_inventory)
+      inventory.create_target_from_plan('uri' => uri)
+    end
+
+    # URI can be passes as nil
+    def initialize(uri = nil,
+                   name = nil,
+                   target_alias = nil,
+                   config = nil,
+                   facts = nil,
+                   vars = nil,
+                   features = nil)
+      @name = name
+    end
+    # rubocop:enable UnusedMethodArgument
+
+    # Used for munging target + group data
+    def target_data_hash
+      {
+        'config' => @inventory.targets[@name]['config'],
+        'vars' => @inventory.targets[@name]['vars'],
+        'facts' => @inventory.targets[@name]['facts'],
+        'features' => @inventory.targets[@name]['features'].to_a,
+        'plugin_hooks' => @inventory.targets[@name]['plugin_hooks'],
+        'name' => @inventory.targets[@name]['name'],
+        'uri' => @inventory.targets[@name]['uri'],
+        'alias' => @inventory.targets[@name]['target_alias']
+      }
+    end
+
+    # Now that these are attributes, I think we need actual methods for them
+    def features
+      @inventory.features(self)
+    end
+
+    def vars
+      @inventory.vars(self)
+    end
+
+    def facts
+      @inventory.facts(self)
+    end
+
+    def to_s
+      # Using this for debugging in plans, TODO: remove commented out debug line
+      # to_h.merge(target_data_hash).inspect
+      safe_name
+    end
+
+    def safe_name
+      @inventory.targets[@name]['safe_name']
+    end
+
+    def to_h
+      options.merge(
+        'name' => name,
+        'uri' => uri,
+        'protocol' => protocol,
+        'user' => user,
+        'password' => password,
+        'host' => host,
+        'port' => port
+      )
+    end
+
+    def host
+      @inventory.targets[@name]['uri_obj']&.hostname || @inventory.targets[@name]['host']
+    end
+
+    attr_reader :name
+
+    def uri
+      @inventory.targets[@name]['uri']
+    end
+
+    def remote?
+      @inventory.targets[@name]['uri_obj']&.scheme == 'remote' || @inventory.targets[@name]['protocol'] == 'remote'
+    end
+
+    def port
+      @inventory.targets[@name]['uri_obj']&.port || @inventory.targets[@name]['port']
+    end
+
+    # transport is separate from protocol for remote targets.
+    def transport
+      remote? ? 'remote' : protocol
+    end
+
+    def protocol
+      @inventory.targets[@name]['uri_obj']&.scheme || @inventory.targets[@name]['protocol']
+    end
+
+    def user
+      unencode(@inventory.targets[@name]['uri_obj']&.user) || @inventory.targets[@name]['user']
+    end
+
+    def password
+      unencode(@inventory.targets[@name]['uri_obj']&.password) || @inventory.targets[@name]['password']
+    end
+
+    def options
```

Comment:
Should this have an alias to `.config`? It seems weird the config is the only object attribute that doesn't have a getter.

Reply:
I think we can expose 'target_config` which would preserve the nested hash structure to allow something like:
```
Target.new('name' => 'copy', 'config' => get_target('blueprint').config)
```

Reply:
Exposed config